### DISCUSSION TITLE
next/link with typescript and passHref

### DISCUSSION THREAD

**Discussion by 0xsven** [vercel/next.js#19822] [Help]
I have a link component like so:

```tsx
import { forwardRef } from "react";

interface Props {
  children: ReactNode;
  href: string;
}

const MyLink = forwardRef<HTMLAnchorElement, Props>(
  ({ children, href }, ref) => (
    <a ref={ref} href={href}>
      {children}
    </a>
  )
);

export default MyLink;
```

And I use it like so:
```tsx
<Link href="/about" passHref>
  <MyAnchor>About</MyAnchor>
</Link>
```

I am getting a type issue: `Property 'href' is missing in type '{ children: string; }' but required in type 'Props'.` 

It works if I change my code to this: 

```tsx
<Link href="/about" passHref>
  <MyAnchor href="/about">About</MyAnchor>
</Link>
```

But this looks wrong to me. How to do it?

(try it yourself here: [codesandbox](https://codesandbox.io/s/still-frost-w8h67?file=/pages/index.tsx))

**Comment by pmrt** [comment]
This is another reason to fix our old friend: #7915 (which is still closed without a solution). In addition to `href`, according to [this](https://nextjs.org/docs/api-reference/next/link#if-the-child-is-a-function-component): it should be something like:

```tsx
import { forwardRef } from "react";

// Add onClick:

interface Props {
  children: ReactNode;
  href: string;
  onClick: MouseEventHandler<HTMLAnchorElement>; // <------------------------- 
}

const MyLink = forwardRef<HTMLAnchorElement, Props>(
  ({ children, href, onClick }, ref) => ( // <------------------------------ 
    <a ref={ref} href={href} onClick={onClick}> // <---------------------- 
      {children}
    </a>
  )
);

export default MyLink;
```

Because as stated in the link above, we need to pass `onClick` too (no more info provided in the docs). So now we have two properties implicitly passed, like a black box, to our components which are not friendly at all with Typescript, one of them needs "passHref" and the other doesn't but both are required according to docs.

While this is fixed I just use optional types, but it is not ideal:

```tsx
interface Props {
  children: ReactNode;
  href?: string;
  onClick?: MouseEventHandler<HTMLAnchorElement>;
}
```

IMO this is just weird, inconsistent design which would be fixed by fixing #7915.

**Reply by nicosampler** [reply]
💯
